controller/service.go

```diff
@@ -161,16 +162,44 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 		lbIngressIPs = append(lbIngressIPs, v1.LoadBalancerIngress{IP: lbIP.String()})
 	}
 	svc.Status.LoadBalancer.Ingress = lbIngressIPs
+	if svc.Annotations == nil {
+		svc.Annotations = make(map[string]string)
+	}
+	svc.Annotations[annotationIPAllocateFromPool] = pool
 	return true
 }
 
 // clearServiceState clears all fields that are actively managed by
 // this controller.
 func (c *controller) clearServiceState(key string, svc *v1.Service) {
 	c.ips.Unassign(key)
+	delete(svc.Annotations, annotationIPAllocateFromPool)
 	svc.Status.LoadBalancer = v1.LoadBalancerStatus{}
 }
 
+// updateService return true if svc's status or annotations changed
+// and update svc. Or nothing to do and return false.
+func (c *controller) updateService(svcRo, svc *v1.Service) bool {
+	isUpdate := false
+	if !reflect.DeepEqual(svcRo.Status, svc.Status) {
```

Comment:
What I don't like about this function is, it says "updateService" but doesn't tell how.
So I'd split it like this

We do the reflect.DeepEqual checks outside, and we split this function in two:
- updateSvcStatus
- updateSvcAnnotations

Also, this function is added as a method of the controller but it's not really using anything from it, I'd convert the two functions to "non methods"

Reply:
Totally agree with you and Thanks. 

- Call `c.client.UpdateStatus` in `updateSvcStatus` and `updateSvcAnnotations` respectively.
- Or call `c.client.UpdateStatus` after calling `updateSvcStatus` and `updateSvcAnnotations`? 

Which is better?  I prefer option 2 , because this only call the k8s api once.

Reply:
Yep, 2 is better. The "real" call can still happen outside

Reply:
Thanks, Done